sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+       PADCTRL_MIO_PADS3_ATTR13_OFFSET, 13},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR14_MASK,
+       PADCTRL_MIO_PADS3_ATTR14_OFFSET, 14},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR15_MASK,
+       PADCTRL_MIO_PADS3_ATTR15_OFFSET, 15},
+      {PADCTRL_MIO_PADS4_REG_OFFSET, PADCTRL_MIO_PADS4_ATTR16_MASK,
+       PADCTRL_MIO_PADS4_ATTR16_OFFSET, 16},
+      {PADCTRL_MIO_PADS4_REG_OFFSET, PADCTRL_MIO_PADS4_ATTR17_MASK,
+       PADCTRL_MIO_PADS4_ATTR17_OFFSET, 17},
+      {PADCTRL_MIO_PADS4_REG_OFFSET, PADCTRL_MIO_PADS4_ATTR18_MASK,
+       PADCTRL_MIO_PADS4_ATTR18_OFFSET, 18},
+      {PADCTRL_MIO_PADS4_REG_OFFSET, PADCTRL_MIO_PADS4_ATTR19_MASK,
+       PADCTRL_MIO_PADS4_ATTR19_OFFSET, 19},
+      {PADCTRL_MIO_PADS5_REG_OFFSET, PADCTRL_MIO_PADS5_ATTR20_MASK,
+       PADCTRL_MIO_PADS5_ATTR20_OFFSET, 20},
+      {PADCTRL_MIO_PADS5_REG_OFFSET, PADCTRL_MIO_PADS5_ATTR21_MASK,
+       PADCTRL_MIO_PADS5_ATTR21_OFFSET, 21},
+      {PADCTRL_MIO_PADS5_REG_OFFSET, PADCTRL_MIO_PADS5_ATTR22_MASK,
+       PADCTRL_MIO_PADS5_ATTR22_OFFSET, 22},
+      {PADCTRL_MIO_PADS5_REG_OFFSET, PADCTRL_MIO_PADS5_ATTR23_MASK,
+       PADCTRL_MIO_PADS5_ATTR23_OFFSET, 23},
+      {PADCTRL_MIO_PADS6_REG_OFFSET, PADCTRL_MIO_PADS6_ATTR24_MASK,
+       PADCTRL_MIO_PADS6_ATTR24_OFFSET, 24},
+      {PADCTRL_MIO_PADS6_REG_OFFSET, PADCTRL_MIO_PADS6_ATTR25_MASK,
+       PADCTRL_MIO_PADS6_ATTR25_OFFSET, 25},
+      {PADCTRL_MIO_PADS6_REG_OFFSET, PADCTRL_MIO_PADS6_ATTR26_MASK,
+       PADCTRL_MIO_PADS6_ATTR26_OFFSET, 26},
+      {PADCTRL_MIO_PADS6_REG_OFFSET, PADCTRL_MIO_PADS6_ATTR27_MASK,
+       PADCTRL_MIO_PADS6_ATTR27_OFFSET, 27},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR28_MASK,
+       PADCTRL_MIO_PADS7_ATTR28_OFFSET, 28},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR29_MASK,
+       PADCTRL_MIO_PADS7_ATTR29_OFFSET, 29},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR30_MASK,
+       PADCTRL_MIO_PADS7_ATTR30_OFFSET, 30},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR31_MASK,
+       PADCTRL_MIO_PADS7_ATTR31_OFFSET, 31}};
+
+  // Vector of all attributes, initialise with the basic attributes.
+  std::vector<dif_padctrl_attr_t> attributes_ = {
+      kDifPadctrlAttrIOInvert, kDifPadctrlAttrOpenDrain,
+      kDifPadctrlAttrPullDown, kDifPadctrlAttrPullUp,
+      kDifPadctrlAttrKeeper,   kDifPadctrlAttrWeakDrive};
+
+  /**
+   * Setup dummy `mmio_region_t` suitable for use with mock testing and
+   * initialise padctrl DIF using it.
+   */
+  void DeviceInit() {
+    mmio_region_t base_addr = dev().region();
+    EXPECT_EQ(dif_padctrl_init(base_addr, &dif_padctrl_), kDifPadctrlOk);
+  }
+
+  /**
+   * Expect a read of the REGEN register to check the lock bit
+   *
+   * @param locked Whether REGEN will report padctrl is locked or not
+   */
+  void ExpectLockCheck(bool locked) {
+    EXPECT_READ32(PADCTRL_REGEN_REG_OFFSET, locked ? 0 : 1);
+  }
```

Comment:
I would recommend against writing really short helpers like these.

`EXPECT` macros use `__LINE__` to give better test error messages, and such helpers effectively break that. This means that `EXPECT` helpers should only be used when absolutely necessary, rather than for single lines like this.

I don't think we have specific guidance on this, but it's typically better to define helpers as macros rather than as functions to help preserve line information.

Reply:
I've spun them out into macros, felt it made the tests more readable rather then using the plain `EXPECT_EQ` and `EXPECT_READ32`